### PR TITLE
[PLTF-1690] Add support for testing with multiple Ruby versions

### PR DESCRIPTION
**Pull request by vicente-ramos**
Perhaps is better to leave gem-test as the target that test against multiple Ruby versions and create (or remove?) a new simple-gem-test target to test against environment default Ruby version.

### CODE REVIEW

File: artifact-management/commands/gem-test-common-steps.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+description: Perform common gem test steps
+parameters:
+  gem_test_dir:
+    description: The directory in which the gem tesets should run
+    type: string
+    default: .
+  bundler_version:
```
Comment (nickmarden) [inline artifact-management/commands/gem-test-common-steps.yml:7]: This default should probably be sensitive to the version of Ruby being used. For Ruby 2.7 or later, Bundler 2.x is already included in the Ruby distribution and it is very hard to remove it.

File: artifact-management/commands/gem-test-multiple-ruby-versions.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    type: string
+    default: .
+  test_command:
+     default: bundle exec rake test
+     description: The command that should be used to run tests
+     type: string
+  ruby_versions:
+    description: The ruby versions to be used to test the gem.
+    type: string
+    default: "2.2.10,2.3.8,2.4.10,2.5.8,2.6.6,2.7.1"
```
Comment (nickmarden) [inline artifact-management/commands/gem-test-multiple-ruby-versions.yml:14]: Although I think it's great that our `avvo/ruby-testing` Docker image has Ruby 2.7.x, by default we probably don't want to test against it unless explicitly requested. Why?

- Ruby 2.7.x uses Bundler 2.0, which may clash with the specified Bundler version in the `Gemfile.lock` file.
- Ruby 2.7.x (and even 2.6.x) include some new behaviors that are likely to conflict with older Ruby code. In other words, the tests probably won't pass in most cases.

File: artifact-management/commands/gem-test.yml [hunk 3]
```diff
@@ -1,36 +1,16 @@
-description: Perform the actual gem build
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:1]: You've made quite a few changes in this file, which are unrelated. They should be in separate commits to tell a cleaner story. Remember that commits tell a story to the developer who reads them later. Make the story as easy as possible to understand.

The change here at line 1 is "correct the description of the `gem-test` command". It's unrelated to the rest of the changes.

File: artifact-management/commands/gem-test.yml [hunk 4]
```diff
@@ -1,36 +1,16 @@
...
+description: Perform simple gem test
 parameters:
-  bundler_version:
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:3]: Here you've removed my ability to specify the `bundler_version`, I think. I realize it's part of the new `gem-test-common-steps` but as a user I'm not going to call `gem-test-common-steps` directly, I'm probably going to call `gem-test` or `gem-test-multiple-ruby-versions`. I want to call it like this:

```
- gem-test:
  bundler_version: 1.17.4
```

and I expect the `bundler_version` to be used wherever it is needed (in this implementation, it would be passed into the `gem-test-common-steps` command. But you've taken that ability away from me by removing this parameter here.

File: artifact-management/commands/gem-test.yml [hunk 5]
```diff
@@ -1,36 +1,16 @@
...
-    description: The Bundler version that should be used for "bundle install" and the actual test command
-    type: string
-    default: 1.17.3
-  gem_repo_host:
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:7]: I think that `gem_repo_host` and `gem_repo_key` were copy-pasta (🍝 ) from a different command, so you're correct that they should be removed. But they should be removed in a separate commit with a message like "Removed unused parameters from gem-test" and not part of this monolithic commit. Again, to tell a cleaner story.

File: artifact-management/commands/gem-test.yml [hunk 6]
```diff
@@ -1,36 +1,16 @@
...
-    default: AVVO_GEM_REPO_HOST
-  gem_repo_key:
-    description: The :rubygems_api_key used to access the gem repository host
-    type: env_var_name
-    default: AVVO_GEM_REPO_TOKEN
   gem_test_dir:
     description: The directory in which the gem tesets should run
     type: string
     default: .
-  packagecloud_read_token:
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:19]: I think that `setup-gem-environment` actually requires the `packagecloud_read_token`. By taking this away you aren't preventing me from specifying the `packagecloud_read_token` because (like you did with `bundler_version`) this setting is specified by environment variables, and the environment variable set in the CircleCI build context will still be visible in the executor process due to Avvo's `org-global` context configuration.

However, what you *have* done here is take away the _documentation_ that this environment variable is going to be consumed by the `gem-test` command, and that's not great. I had it listed here for that exact reason.

File: artifact-management/commands/gem-test.yml [hunk 7]
```diff
@@ -1,36 +1,16 @@
...
     type: string
-    default: bundle exec rake test
 steps:
   - setup-gem-environment
-  - run:
-      name: Install bundler gem
-      command: |-
-        cd <<parameters.gem_test_dir>>
-        gem install bundler --version=<<parameters.bundler_version>>
+  - gem-test-common-steps
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:13]: TODO: Figure out how to pass the `bundler_version` that should be specifiable in `gem-test` into the `gem-test-common-steps` subcommand.

File: artifact-management/commands/gem-test-common-steps.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+description: Perform common gem test steps
+parameters:
+  gem_test_dir:
+    description: The directory in which the gem tesets should run
```
Comment (nickmarden) [inline artifact-management/commands/gem-test-common-steps.yml:4]: `tesets` is a typo (on my part I think) and should be fixed.

**Comment by vicente-ramos** [comment]
gem-test target "is smarter" in the sense that have two main behavior:

- Legacy behavior: Already existing code in .circleci/config.yml keeps working as it was. As if no change were introduced.

- If ruby_versions parameter is specified to gem-test target the rvm command will be used (no matter if one or more versions are specified). An image with rvm must be used (as avvo/ruby-testing)

Will this new gem-test target behavior require a README.md update to document it?

File: artifact-management/commands/gem-test.yml [hunk 9]
```diff
@@ -1,44 +1,87 @@
-description: Perform the actual gem build
+description: Perform gem test with default only or many Ruby versions
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:1]: Edit for clarity:

> Perform gem test with system Ruby, or with one or more Ruby versions (using RVM) if ruby_versions is set. ruby_versions should be a comma-separated string with no spaces, such as "2.2.10,2.3.8,2.5.8"

File: artifact-management/commands/gem-test.yml [hunk 10]
```diff
@@ -1,44 +1,87 @@
...
+    description: The token to be used when fetching packages from Packagecloud.
     type: env_var_name
-    default: AVVO_GEM_REPO_HOST
+    default: PACKAGECLOUD_READ_TOKEN
   gem_repo_key:
     description: The :rubygems_api_key used to access the gem repository host
     type: env_var_name
     default: AVVO_GEM_REPO_TOKEN
+  bundler_1_version:
+    description: The Bundler version that should be used for "bundle install" and the actual test command
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:12]: > The precise Bundler 1.x version

File: artifact-management/commands/gem-test.yml [hunk 11]
```diff
@@ -1,44 +1,87 @@
...
+    type: string
+    default: 1.17.3
+  bundler_2_version:
+    description: The Bundler version that should be used for "bundle install" and the actual test command
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:16]: > The precise Bundler 2.x version

File: artifact-management/commands/setup-gem-environment.yml [hunk 12]
```diff
@@ -4,10 +4,6 @@ parameters:
     description: The token to be used when fetching packages from Packagecloud.
     type: env_var_name
     default: PACKAGECLOUD_READ_TOKEN
-  gem_repo_host:
```
Comment (nickmarden) [inline artifact-management/commands/setup-gem-environment.yml:7]: Check me on this, but I *think* that we only use `gem_repo_host` and `gem_repo_key` for pushes, not for pulls. So either we need both of these parameters or we need neither of them.

But please check this, I could be wrong.

The same comment applies to [the place where you removed `gem_repo_host`](https://github.com/avvo/circleci-orbs/pull/6/files#diff-6e9e49cb9a90e06b4611ca23c6a0fd20L7) in the `gem-test` command.

**Review by nickmarden — CHANGES_REQUESTED** [review]

File: artifact-management/commands/gem-test.yml [hunk 13]
```diff
@@ -1,44 +1,87 @@
...
-        <<parameters.test_command>>
+  - when:
+      condition: <<parameters.ruby_versions>>
+      steps:
+        - run:
+            name: Install bundler gem
+            command: |-
+              for v in $(echo "<<parameters.ruby_versions>>" | sed "s/,/ /g");
+              do \
+                if [[ "$v" > "2.6.9999" ]];
```
[the remark this answers is not in the document]
Comment (nickmarden) [reply artifact-management/commands/gem-test.yml:41]: Additional tests:

```
$ if (( 17 > 3 )); then echo bigger; else echo smaller; fi
bigger
$ if (( 17.1 > 17.0 )); then echo bigger; else echo smaller; fi
bigger
$ if (( 17.1 > 17.2 )); then echo bigger; else echo smaller; fi
smaller
$ if (( "17.1" > "17.2" )); then echo bigger; else echo smaller; fi
zsh: bad math expression: illegal character: "
smaller
$ if (( 17.1 > 17.2 )); then echo bigger; else echo smaller; fi 
smaller
$ v1=17.1
$ v2=17.2
$ if (( $v1 > $v2 )); then echo bigger; else echo smaller; fi
smaller
$ v1=17.3                                                    
$ if (( $v1 > $v2 )); then echo bigger; else echo smaller; fi
bigger
```

[the remark this answers is not in the document]
Comment (nickmarden) [reply artifact-management/commands/gem-test.yml:41]: Also please note: **you're assuming these tests will run under bash**. Is it always the case that `gem-test` runs under bash? Be very careful about this assumption because it depends on which docker image is being used to run the tests, I think.

[the remark this answers is not in the document]
Comment (vicente-ramos) [reply artifact-management/commands/gem-test.yml:41]: From https://github.com/CircleCI-Public/slack-orb/blob/staging/src/commands/approval.yml#L48-L54 I take the idea of giving an error if there is no bash, this is because I could not found how to “require bash”. Then for both, when branch and unless branch, is specifed that /bin/bash must be used as shell.

File: artifact-management/commands/gem-test.yml [hunk 14]
```diff
@@ -1,44 +1,87 @@
...
+              rvm <<parameters.ruby_versions>> do <<parameters.test_command>>
+  - unless:
+     condition: <<parameters.ruby_versions>>
+     steps:
+       - run:
+           name: Install bundler gem
+           command: |-
+             ACTUAL_RUBY=$(ruby --version);
+             RUBY_VERSION=${ACTUAL_RUBY:5:3};
+             if [[ "$RUBY_VERSION" > "2.6.9999" ]]
```
[the remark this answers is not in the document]
Comment (vicente-ramos) [reply artifact-management/commands/gem-test.yml:68]: In the link https://tldp.org/LDP/abs/html/comparison-ops.html says:
>
```
is greater than, in ASCII alphabetical order
if [[ "$a" > "$b" ]]
if [ "$a" \> "$b" ]
Note that the ">" needs to be escaped within a [ ] construct.
See Example 27-11 for an application of this comparison operator.
```
Then using [ ] constructor (and escape “>”) to standardize with other gems in same orb (like setup-gem-environment)

File: artifact-management/commands/gem-test.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (nickmarden) [reply artifact-management/commands/gem-test.yml]: Yes, please use this approach.

I believe you only need to check for bash when you are using the sophisticated numeric comparison operators we are discussing elsewhere in this PR conversation, but I don't have a problem with it being a  universal requirement for all of the build upgrades we will be doing.

File: artifact-management/commands/gem-test.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (nickmarden) [reply artifact-management/commands/gem-test.yml]: I think we can use the `[[ ]]` construct (if we require bash - see the previous conversation about this), but I think you're missing the point:

```
if [[ 17 > 3 ]]; then
  echo bigger
else
  echo smaller
fi

smaller
```

**That result is wrong** because you're comparing in alphabetical order. 17 is greater than 3, not less than. You need to require bash and then use `(( ))` instead of `[[ ]]` to get _numeric_ comparison, which is what you are looking for.

File: artifact-management/commands/setup-gem-environment.yml (hunk 12 above)
Comment (nickmarden) [reply artifact-management/commands/setup-gem-environment.yml:7]: We're missing each other a bit on this one. Now that I've re-read my comments I can see why my point isn't clear. Let me try again.

I think that `bundle install` or `gem install` will _pull_ from artifactory without the `gem_repo_key` written to the `$HOME/.gem/credentials` file. I think that file is only used for _pushes_.

So in other words `setup-gem-environment` is configuring for two cases:

* One case where we are pulling from Packagecloud (credentials required) and Artifactory (credentials not required)
* One case where we are pushed to Packagecloud (credentials required) and Artifactory (credentials required)

We use `setup-gem-environment` in both cases, because it installs the maximum number of credentials required.

Is it necessary to separate this into `setup-gem-pull-environment` and `setup-gem-push-environment`? Probably not. It's correct but also somewhat pedantic.

I think that the removal of `gem_repo_host` without removing `gem_repo_key` - as you originally proposed - is probably the simplest way forward.

File: artifact-management/commands/gem-test.yml [hunk 15]
```diff
@@ -1,33 +1,35 @@
...
-    type: env_var_name
-    default: AVVO_GEM_REPO_TOKEN
   bundler_1_version:
-    description: The Bundler version that should be used for "bundle install" and the actual test command
+    description: The precise Bundler 1.x version
     type: string
     default: 1.17.3
   bundler_2_version:
-    description: The Bundler version that should be used for "bundle install" and the actual test command
+    description: The precise Bundler 2.x version
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:10]: I meant "The precise Bundler 2.x version _that should be used..._". The rest of the sentence was important! I was just providing an edit to the (ambiguous phrase) "The Bundler version" which should instead be "The precise Bundler 1.x version" or "The precise Bundler 2.x version"

File: artifact-management/commands/gem-test.yml [hunk 16]
```diff
@@ -1,44 +1,98 @@
...
     type: string
     default: .
   packagecloud_read_token:
     description: The token to be used when fetching packages from Packagecloud.
     type: env_var_name
     default: PACKAGECLOUD_READ_TOKEN
-  test_command:
-    description: The command that should be used to run tests
+  ruby_versions:
+    description: The ruby versions to be used to test the gem.
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:26]: Add to this description: "If set, RVM is assumed to be installed and the install/test commands will be prefaced with `rvm <<ruby_versions> do ...`. If unset (the default), the system ruby will be called directly."

File: artifact-management/commands/gem-test.yml [hunk 17]
```diff
@@ -1,44 +1,98 @@
...
+  - when:
+      condition: <<parameters.ruby_versions>>
+      steps:
+        - run:
+            name: Install bundler gem
+            shell: /bin/bash
+            command: |-
+              for v in $(echo "<<parameters.ruby_versions>>" | sed "s/,/ /g");
+              do \
+                if [ "$v" \> "2.6.9999" ];
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:51]: I think this should be `if (( $v > 2.6.9999 ))` per the prior discussion, right?

File: artifact-management/commands/gem-test.yml [hunk 18]
```diff
@@ -1,44 +1,98 @@
...
+  - unless:
+     condition: <<parameters.ruby_versions>>
+     steps:
+       - run:
+           name: Install bundler gem
+           shell: /bin/bash
+           command: |-
+             ACTUAL_RUBY=$(ruby --version);
+             RUBY_VERSION=${ACTUAL_RUBY:5:3};
+             if [ "$RUBY_VERSION" \> "2.6.9999" ]
```
Comment (nickmarden) [inline artifact-management/commands/gem-test.yml:79]: Likewise here with `(( ))`

File: artifact-management/commands/gem-test.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (vicente-ramos) [reply artifact-management/commands/gem-test.yml]: 😳 I apologize about this rude mistake. No excuse. 
New proposed solution should work in any Bourne shell; but a cons about it is that it assumes ruby 1.x is never used. What you think?

File: artifact-management/commands/setup-gem-environment.yml (hunk 12 above)
Comment (vicente-ramos) [reply artifact-management/commands/setup-gem-environment.yml:7]: I inclined to think that we should hear DRY philosophy for this case. Despite that setup-gem-environmen is configuring a case that is not required when pulling (Artifactory) I found better to have all “credential” logic in one single place.

Comment (vicente-ramos) [reply artifact-management/commands/setup-gem-environment.yml:7]: Other issue about this variables is:
But what I do not understand is why we are using this code like next:
```
- run:
    name: "Configure destination gem repo"
    command: |-
      printf -- "---\n:rubygems_api_key: Basic $AVVO_GEM_REPO_TOKEN\n" > $HOME/.gem/credentials
      chmod 0600 $HOME/.gem/credentials
```
	where there is no need to declare the parameter `gem_repo_key` (of type `env_var_name`) because is not used in the code.
	Instead of using a code more like suggested here: [https://circleci.com/docs/2.0/reusing-config/#environment-variable-name](Environment Variable Name) That it should look something like:
```
- run:
    name: "Configure destination gem repo"
    command: |-
      printf -- "---\n:rubygems_api_key: Basic **${<<parameters.gem_repo_key>>}**\n" > $HOME/.gem/credentials
      chmod 0600 $HOME/.gem/credentials
```
	Could the answer be something like?: Because we are not expecting the users to give a new value to this variables: we do not expect a use like:
```
jobs:
  test:
        docker:
      - image: circleci/ruby:2.6
    steps:
      - checkout
      - artifact-management/gem-test:
            gem_repo_key: other_value_for_this
```
	But if this is the case, why are we declaring the parameter `gem_repo_key`?

Comment (nickmarden) [reply artifact-management/commands/setup-gem-environment.yml:7]: We're declaring it for documentation, so that readers of the orb docs realize what effect the environment variable `AVVO_GEM_REPO_TOKEN` has on the commands that consume it.

We can implement the command script using either `$AVVO_GEM_REPO_TOKEN` or `<<parameters.gem_repo_key>>`; it works the same either way. However I don't think you can override an `env_var_name` parameter in an orb command call like you hypothesized in your example. So, just to reiterate, the purpose is documentation, not overridability.

File: artifact-management/commands/gem-test.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (nickmarden) [reply artifact-management/commands/gem-test.yml]: Yeah, I think this is good enough. We don't use Ruby 1.x anywhere.

Of course, it's possible to write a compound predicate that considers the case where `(major >=3) || (major == 2 && minor >=7)` but I'm not sure it's worth it.

Also, `mayor` in Spanish is spelled `major` in English. But everyone knows what you meant.

**Review by nickmarden — APPROVED** [review]
